Fix CV1 GridItemsLayout centering single item AND Fix Empty view not resizing when bounds change

albyrock87 opened the pull request:
### Description of Change

`GridViewLayout` has a special positioning override when there's only one item.

```
if (ScrollDirection == UICollectionViewScrollDirection.Vertical && NeedsSingleItemHorizontalAlignmentAdjustment(layoutAttributesForRectElements))
{
	// If there's exactly one item in a vertically scrolling grid, for some reason UICollectionViewFlowLayout
	// tries to center it. This corrects that issue.
	var currentFrame = layoutAttributesForRectElements[0].Frame;
	var newFrame = new CGRect(CollectionView.Frame.Left + CollectionView.ContentInset.Right,
	currentFrame.Top, currentFrame.Width, currentFrame.Height);
	layoutAttributesForRectElements[0].Frame = newFrame;
}
```

So we have to invalidate the layout entirely to trigger that special case, instead of just simply invalidate single cells.

### Issues Fixed

<!-- Please make sure that there is a bug logged for the issue being fixed. The bug should describe the problem and how to reproduce it. -->

Fixes #29595
Fixes #29634

<!--
Are you targeting main? All PRs should target the main branch unless otherwise noted.
-->

jsuarezruiz commented:
/azp run

rmarinho reviewed (changes requested):
I think we just need to update the screenshot.

albyrock87 commented on src/Controls/tests/TestCases.Shared.Tests/Tests/Issues/Issue29595_29634.cs:
```diff
@@ -0,0 +1,24 @@
+	{
+		// This verifies both 29595 and 29634
+		App.WaitForElement("StubLabel");
+		VerifyScreenshot();
```
@jsuarezruiz done, thanks!

rmarinho commented:
/azp run MAUI-UITests-public

rmarinho commented:
/azp run MAUI-UITests-public

mattleibow commented on src/Controls/src/Core/Handlers/Items/iOS/ItemsViewController.cs:
```diff
@@ -239,24 +243,32 @@ private protected virtual void LayoutSupplementaryViews()
-				CollectionView.CollectionViewLayout.InvalidateLayout(layoutInvalidationContext);
+				// GridLayout has a special positioning override when there's only one item
+				// so we have to invalidate the layout entirely to trigger that special case.
+				if (ItemsSource.ItemCount == 1)
+				{
+					CollectionView.CollectionViewLayout.InvalidateLayout();
+				}
```
Is this extra code path not needed in CV2?

mattleibow commented on src/Controls/tests/TestCases.HostApp/Issues/Issue29595_29634.cs:
```diff
@@ -0,0 +1,366 @@
+using System.Windows.Input;
+using Microsoft.Maui.Controls.Shapes;
+
+namespace Maui.Controls.Sample.Issues
```
This test is really complex and there are many things here. Does this issue not repro in a case without all of this special work?

I see multiple collection views, shell, pages, tabs and all sorts of delay things.

Can this not be broken into a simple test where the page loads and then the UI test taps a button to add an item if we need things loaded later.

Also, the fix for the empty view has no tests I think?

mattleibow reviewed (changes requested):
Thanks for the PR, I have a few questions:

* I am not seeing a visible difference in the final logic of the CV2 code
* Similar with CV1 - except for a final if and layout all
* The tests is very complex and I am not sure what we are testing
* The PR has a fix for 2 things, but there is only 1 test
* The CV code has logic for horizontal and vertical grids, should we add tests for each
* Does this just manifest if the items change after the page loads, or also on the page init? Should we cover both cases?

albyrock87 committed:
Fix CV1 empty view not resizing when bounds change and grid layout not left aligning single item

albyrock87 commented:
@mattleibow I should have addressed all your concerns in the last force-push

- I've simplified/separated the UITests (they were initially done based on user reports while now they focus on the single issues)
- CV2 is not affected by those bugs because:
  - Supplementary views are handled differently (not hacky)
  - It uses the compositional layout so the grid layout left-align on single item is not hacky

PureWeen commented on src/Controls/tests/TestCases.HostApp/Issues/Issue29634.cs:
```diff
@@ -0,0 +1,40 @@
+﻿namespace Maui.Controls.Sample.Issues;
+
+[Issue(IssueTracker.Github, 29634, "iOS CV: Empty view not resizing when bounds change", PlatformAffected.iOS)]
+public class Issue29634 : ContentPage
```
If I revert the changes on this PR this test still passes

PureWeen committed:
- rework test

PureWeen committed:
- fix test

PureWeen commented:
/backport to release/9.0.1xx-sr8
